src/zarr/buffer.py

```diff
@@ -261,6 +261,11 @@ def __add__(self, other: Buffer) -> Self:
             np.concatenate((np.asanyarray(self._data), np.asanyarray(other_array)))
         )
 
+    def __eq__(self, other: object) -> bool:
+        # Note: this was needed to support comparing MemoryStore instances with Buffer values in them
+        # if/when we stopped putting buffers in memory stores, this can be removed
+        return isinstance(other, type(self)) and self.to_bytes() == other.to_bytes()
```

Comment:
for reference, because I had to look this up: `and` short-circuits, so if the first leg of the operator is `False` we never invoke the potentially expensive second leg, which is good.

Reply:
Can we make this a separate utility function that compares a buffer with any other arbitrary object?
```python
class Buffer:
    ...
    def __eq__(self, other: Buffer) -> bool:
        # Another Buffer class can override this to choose a more efficient path
        return np.allclose(self._data, other.as_numpy_array())

def buffers_equal(buffer: Buffer, other: object) -> bool:
    return isinstance(other, Buffer) and buffer == other
```

Reply:
@akshaysubr - happy to. Your example makes mypy unhappy though:

```
src/zarr/core/buffer.py:267: error: Argument 1 to "allclose" has incompatible type "ArrayLike"; expected "_SupportsArray[dtype[Any]] | _NestedSequence[_SupportsArray[dtype[Any]]] | bool | int | float | complex | str | bytes | _NestedSequence[bool | int | float | complex | str | bytes]"  [arg-type]
src/zarr/core/buffer.py:267: error: "object" has no attribute "as_numpy_array"  [attr-defined]
```
Would this be okay?

```python
    def __eq__(self, other: object) -> bool:
        # Another Buffer class can override this to choose a more efficient path
        try:
            return np.allclose(self._data, other.as_numpy_array())  # type: ignore[arg-type, attr-defined]
        except AttributeError:
            return False
```

Reply:
It might be cleaner to just do something like this so both arguments are guaranteed to be numpy arrays?
```python
    def __eq__(self, other: Buffer) -> bool:
        # Another Buffer class can override this to choose a more efficient path
        return np.array_equal(self.as_numpy_array(), other.as_numpy_array())
```